### DISCUSSION TITLE
Custom DDP messages

### DISCUSSION THREAD

**Discussion by satyavh** [meteor/meteor#11522] [Feature Requests]
Meteor provides DDP over Sockjs but there's weirdly no easy way (without hacking and diving deep into core code) to just send messages around from server <-> clients using the existing socket / ddp connection. You basically have to setup a new sockjs/socket.io connection to accomplish this, or use a database backed method using pub-sub.

There have been attempts that work, but it's not clear if they are reliable. A reliable solution provided in core would be a great addition to Meteor's real-time capabilities.

2 use cases as an example:

### 1. File upload progress
1. user is uploading something via a method, let's say an array of 200 items.
2. the server is processing the array, taking about 1 second per item.
3. the user wants to see upload progress.

### 2. Worker progress
1. the user triggers a worker by clicking a button
2. the worker does some heavy calculation, it takes couple of minutes
3. the user can keep using the app, but a status is shown of the workers' progress

### 3. Channels
1. 5 users register with a message channel
2. the server and each user can now send a message to the other users via that channel
3. perhaps the server pings all the users that a file or worker is done

Of course it can all be achieved using publications, but then the database is involved which is only a viable option if the data needs to be persisted. Method polling is another option but overkill. 

It's not like this is rocket science either, Meteor already has socket / ddp connection. Just provide a way to natively use those connection. [It can be done](https://github.com/wojtkowiak/meteor-custom-protocol) via hacks, but it would be better if it was supported in core.

**Comment by evolross** [comment]
Agreed. We use [redis-oplog's Vent](https://github.com/cult-of-coders/redis-oplog/blob/master/docs/vent.md) now. But it requires a Redis server.

**Reply by afrokick** [reply]
Vent is useful when you want to send the message to all clients on all servers(processes).

Maybe for room-based games where all players connected to the same server, the custom DDP messages more suitable than broadcasting via redis.

**Comment by filipenevola** [comment]
Hi @satyavh I believe the idea in general is good but your discussion description is lacking details. You can read [here](https://github.com/meteor/meteor/blob/devel/CONTRIBUTING.md#feature-requests) what we would expect.

In your specific case I believe first it would be nice to describe a few real world examples.

Like, I had an app where I had to send X to the server and receive back Y and it was not enough to use Methods or Publications because of W and Z. So it would be nice to have the ability to send messages like M etc etc.

This would help us to review and provide better insights for the solution.

After that we would expect a general idea of the solution, for example, if Meteor could accept any DDP message and then maybe with a custom field specifying that this message is not handled by Meteor core itself you could create a package to communicate in a reliable way. Maybe just allowing any message to go back and forth is enough to open the doors for solutions inside packages that not necessarily need to belong to the core.

Anyway, I would like to see more details about the problem, use cases and also the solution so the chances of implementing your suggestions will be higher.

If you need help for anything that I'm explaining above feel free to make that clear here as well so we can work together.

**Reply by satyavh** [reply]
Thanks for the feedback! I added some examples that I encounter, but I'm sure other devs have other requirements.
